### ISSUE TITLE
Find further metrics to consider

### ISSUE THREAD

**Issue by eddableheath** [alan-turing-institute/ARC-LoCoMoSeT#49]
With the decision to drop the TDA approach, and with PARC currently in question, we are left with only one task specific metric (Renggli). We should find some new candidates to test.

**Comment by eddableheath** [comment]
Further metrics to consider:

- [LogME](https://arxiv.org/pdf/2102.11005.pdf): also found [here](https://arxiv.org/pdf/2110.10545.pdf), with a repo implementing it (among other metrics) [here](https://github.com/thuml/LogME/blob/main/LogME.py). Looks to be fairly accurate according to the results. Complicated algorithm, however has an implementation available to reference.
- [H_alpha](https://openreview.net/pdf?id=iz_Wwmfquno): updates an early metric for this problem to be more effective, claims to be faster than LogME for same accuracy. The paper also suggests problems with NCE and LEEP.
- [NCE](https://openaccess.thecvf.com/content_ICCV_2019/papers/Tran_Transferability_and_Hardness_of_Supervised_Classification_Tasks_ICCV_2019_paper.pdf): appears in a lot of survey papers on this topic, though with varying reports on how effective it is. There are, however, several implementations of it available so shouldn't be too hard to test.
- [LEEP](http://proceedings.mlr.press/v119/nguyen20b/nguyen20b.pdf): and variants, similarly to NCE, appear frequently in the literature with different papers taking different views of them. Also similar to NCE there are implementations available already so testing this shouldn't be a problem.

**Comment by eddableheath** [comment]
Rather than fully implement these into our pipeline one at a time I'll likely compare them with the ImageNet sanity check in one notebook initially for the sake of time and complexity.

**Comment by eddableheath** [comment]
Some notes as I work on this:

For the original 5 models we're looking at LogME doesn't seem to return a good ranking for a small test set, now trying successively larger test sets.

**Comment by eddableheath** [comment]
NCE and LEEP want predicted labels, so running into the same problem as with TDA in that if we don't train a classification head for the new task then we are left using the predicted labels from the original model with a new dataset (which might give a semi reasonable result if they are examining the class separation) or an untrained head (which is questionable if it would work at all). I'll try it for Imagenet as is, and if they give very good results then we can see how that works with either o the above methods.

**Comment by eddableheath** [comment]
On a closer inspection, for a small test set LogME returns a similar ranking to the task-agnostic approach.

**Comment by eddableheath** [comment]
![LogME_4000](https://github.com/alan-turing-institute/ARC-LoCoMoSeT/assets/8882079/aee35531-951f-4f3f-86ae-b476ba2fc8b0)

At 4000 datapoints LogME returns the ranking we want. It's unclear from the paper or the implementation how many datapoints are typically used in the computation of LogME, which suggests to me that they're using the entire dataset. It would be good to have an idea of what the minimum sized probe set would be as inference can be expensive, so more testing is required. I would imagine that this would be related to the number of classes in the dataset. This would appear to be a good choice to experiment with however.

For a smaller probe set size this doesn't work very well either, so it's worth checking that this isn't just a fluke by doing a more extensive test with random subsets of the dataset at more probeset sizes.

**Comment by eddableheath** [comment]
![NCE_500](https://github.com/alan-turing-institute/ARC-LoCoMoSeT/assets/8882079/dd201491-ee85-4fe0-82ea-de3b674fc3f8)


NCE at 500 datapoints also starts to give a good ranking. Worth running similar extended experiments as with LogME to check that this isn't just a one off.

**Comment by eddableheath** [comment]
To clarify the above graphs as they're lacking a lot of information, it's a plot of the Imagenet-1k validation scores for the models (x axis) agains the metric score (y axis) for a subset of imagenet of the size stated in the title of the plots, with the positive correlation between the two being the characteristic that we're after.

**Comment by eddableheath** [comment]
![diet_tiny_probe_set_vs_logme](https://github.com/alan-turing-institute/ARC-LoCoMoSeT/assets/8882079/7942a978-368a-4c5f-ba38-bd7d224ca878)

Looks like the values converge, with different random subsets of imagenet giving similar values.

**Comment by eddableheath** [comment]
![deit_tiny_probe_set_vs_NCE](https://github.com/alan-turing-institute/ARC-LoCoMoSeT/assets/8882079/d550d24d-71f1-4714-80ae-643cdd0f0729)

For NCE the metric value doesn't appear to converge (at least up to the size of probe set currently considered). A better estimation of this might be how the correlation between the ranking returned and the imagenet validation ranking changes as a function of the size of the probe set.

**Comment by eddableheath** [comment]
Todo:

- [x] Implement scripts for both LogME and NCE to fit within our current framework.
- [ ] Run larger experiments on both to assess robustness under variation and the size of the probe set as a hyperparamter.
- [ ] Test NCE for untrained/base-model classification head for new dataset against fine-tuned validation for that dataset (either our own or found within the literature).

**Comment by eddableheath** [comment]
I've misunderstood NCE (but ironically used it correctly in this case, so essentially minterpreted the results):

As an input NCE takes (predicted labels of original dataset from model, labels from new dataset), essentially comparing statistical features (the entropy) of the labels in both cases. As a comparison of ImageNet to ImageNet this is not tremendously insightful as one would expect them to be similar, with the main difference being the accuracy in predicted the labels of each model (i.e. exactly the validation accuracy).

The original paper also considers the task in which the data is the same but the labels vary, however other instances within the literature use it as a point of comparison for the same task as us. 

I'm going to put the NCE research on hold for now, picking it back up when we've done some fine-tuning.

**Comment by eddableheath** [comment]
Todo in order to close this:

- [ ] Test LEEP for sanity check.
- [ ] Implement LEEP script if it passes.
- [ ] Write tests for NCE (and maybe LEEP).
- [ ] Move notebook to notebooks branch.
- [ ] Adapt run script for NCE and LEEP to take advantage of the classes (maybe just LEEP for now if it passes and NCE has got other caveats).
- [ ] Create PR.

**Comment by eddableheath** [comment]
Note LEEP considers entire dataset as standard, but will likely work with some probeset that is proportional (in some sense) to the number of classes.

Also to clarify how LEEP works:

It runs the new dataset images through the pretrained model as it, extracting the final distribution over the source data set (i.e. imagenet) labels. The conditional probability of the target label for an image given this distribution is then estimated (using an emperical maximum likelihood estimator), from which the score is derived.

**Comment by eddableheath** [comment]
It looks like LEEP fails if you don't have an example of each class in the new dataset (fine for smaller datasets but imagenet is going to require a big run, which I'm trying now)

**Comment by eddableheath** [comment]
Currently can't seem to get LEEP to return me a non NaN value, I think it's for the above reason but not entirely sure. If I can't get it fully working I'll make a new issue for it.

**Comment by eddableheath** [comment]
H_alpha-score now covered by #55

**Comment by eddableheath** [comment]
On further inspection I was using LEEP incorrectly (wasn't applying the softmax to normalise the values to a distribution... Clearly the implementation I found requires somewhere it sums to one/is all positive or it throws a divide by zero error!), and I'm now testing it. It looks promising on paper and if I get some reasonable results (looks similar to LogME in that it requires the entire dataset, or at least a large probe set) it won't be hard to add this to out implemenation at this stage.

**Comment by eddableheath** [comment]
Closed by merge.
